api/system/examples/update-state-in-serial-message-callback.rs

```diff
@@ -0,0 +1,68 @@
+#![no_std]
+extern crate alloc;
+
+#[macro_use]
+extern crate sys;
+extern crate playdate_system as system;
+
+use alloc::string::String;
+use core::ptr::NonNull;
+
+use sys::EventLoopCtrl;
+use sys::ffi::*;
+use system::System;
+use system::event::SystemEventExt as _;
+use system::prelude::*;
+
+struct State {
+	latest_message: Option<String>,
+}
+
+impl State {
+	fn new() -> Self { Self { latest_message: None } }
+
+	/// System event handler
+	fn event(&'static mut self, event: SystemEvent) -> EventLoopCtrl {
+		match event {
+			SystemEvent::Init => {
+				System::Default().set_serial_message_callback(Some(|msg| {
+					                 self.latest_message = Some(msg);
+				                 }));
+
+				println!("Game init complete");
+			},
+			_ => {},
+		}
+		EventLoopCtrl::Continue
+	}
+}
+
+impl Update for State {
+	fn update(&mut self) -> UpdateCtrl {
+		if let Some(latest_message) = self.latest_message.clone() {
+			println!("Latest message: {:#?}", latest_message);
+			self.latest_message = None;
+		}
+		UpdateCtrl::Continue
+	}
+}
+
+/// Entry point
+#[no_mangle]
+pub fn event_handler(_api: NonNull<PlaydateAPI>, event: SystemEvent, _sim_key_code: u32) -> EventLoopCtrl {
+	pub static mut STATE: Option<State> = None;
+
+	if unsafe { STATE.is_none() } {
+		let state = State::new();
+		unsafe { STATE = Some(state) }
+	}
+
+	if event == SystemEvent::Init {
+		unsafe { STATE.as_mut().expect("impossible").set_update_handler() }
```

Comment:
I suspect that `set_serial_message_callback` would need to be called here as well to avoid similar issues with trying to mutate `self` within `event`.

After that, I think as long as `State` is structured so that individual fields can be borrowed rather than all of `self` it should be OK.

Reply:
I suppose in real world we have to split the state or wrap it into something like `RefCell`, probably with `Pin`. It costs almost nothing.
This example is just an example that demonstrates how that possible. So that's ok for now I think. 👍🏻

Reply:
> I suppose in real world we have to split the state

I tried something along these lines at first but it didn’t help.

I think I had a separate `Updater` strict that implemented `Update`, which allowed me to set `latest_message` in the closure passed to `set_serial_message_callback` (inside of `event`) but I could no longer access it in `update`.

What did you have in mind?